File: fs-core/src/main/java/xyz/srclab/common/io/ByteBufferOutputStream.java
package xyz.srclab.common.io;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;

final class ByteBufferOutputStream extends OutputStream {

    private final ByteBuffer buffer;

    ByteBufferOutputStream(ByteBuffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
        try {
            buffer.put(b, off, len);
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    @Override
    public void write(byte[] b) throws IOException {
        write(b, 0, b.length);
    }

    @Override
    public void write(int b) throws IOException {
        try {
            buffer.put((byte) b);
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    @Override
    public void flush() throws IOException {
    }

    @Override
    public void close() throws IOException {
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/io/ByteBufferOutputStream.kt
package xyz.srclab.common.io

import java.io.OutputStream
import java.nio.ByteBuffer

/**
 * Makes [ByteBuffer] as [OutputStream].
 */
open class ByteBufferOutputStream(
    private val source: ByteBuffer
) : OutputStream() {

    override fun write(b: Int) {
        source.put(b.toByte())
    }

    override fun write(b: ByteArray) {
        source.put(b)
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
        source.put(b, off, len)
    }
}